### DISCUSSION TITLE
Aggregation of directives from multiple files - yay or nay?

### DISCUSSION THREAD

**Discussion by przemek-pokrywka** [VirtusLab/scala-cli#308] [Using directives and cmd configuration options]
[The docs](https://scala-cli.virtuslab.org/docs/guides/using-directives#why-not-dedicated-configuration-file) state that 
> Using directives can be placed in any .scala files

while also acknowledging that
> configuration distributed across many source files may be hard to maintain in the long term

The proposed solution is to 
> introduce set of lints to ensure that above given project size or complexity all configuration will be centralized

In my opinion, the proposed solution is not sufficient. I think it would be better for the users, that they are aware of the project-wise nature of the directives early on. Ideally, directives would live at the program entry point **or** in the separate `scala.conf` (`program-name.scala-conf`?) file (I'm not in favor of putting uncommented directives into scala files, because of tooling ecosystem incompatibility). I think that directives distributed across many files should be reported as an error.

The reasons against allowing people to distribute directives:
- it will not make it easy for users to upgrade to a proper build tool
- it will allow users to create a mess, without any clear advantages in terms of features
- for [people who like to keep multiple unrelated scripts in a single directory](https://github.com/VirtusLab/scala-cli/issues/260), directives centralized in the program's entry point seem to be the only way to go

**Comment by gurghet** [comment]
I just wanted to bring forward my use case which consists of a fixed number of files (2 right now) for a scaffolding of a cookie-cutter project. Having directives in both files helps a lot since the files have two different owners and one of them only has one directive which is unlikely to clash with the other user file. This two file setup might be useful in cases where the input of a system is a program, like an interview take-home problem or an online teaching platform.

**Reply by SethTisue** [reply]
Perhaps this could error by default but be supported via an opt-in flag? The flag would be highly discoverable, since the error message would presumably mention it.

**Reply by gurghet** [reply]
Maybe we are talking about different things, but the one you describe is the status quo, and my use case is currently very well supported. In fact, the flag is called `--suppress-directives-in-multiple-files-warning`.

**Comment by tgodzik** [comment]
I think we should discourage using directives multiple files as much as possible unless a user knows what they are doing. Setting `--suppress-directives-in-multiple-files-warning` or the related config option seems to indicate that. Other option would be to do it an error even, but that might make it even more irritating for user. Another solution I can think of is putting the config to turn off the warning under `--power` to further limit the amount of users that would use it.

Anyway, for sure this might cause a lot of pain for some users, so a warning is a minimum we can do. We also want to create a command to migrate everything into single file as automatic solution. Do let us know if you feel like we can do more or you have some ideas about how to make it safer in the long run.
